### PR TITLE
Prune unnest mappings using lambda expressions

### PR DESCRIPTION
**Pull request by phd3**
(For enhancement discussed in #3925)

Accomplishes the following two things by transforming an array of rows before UNNEST operation:

1) Remove unreferenced output symbols (eg. `B_BIGINT` in the example below)
2) Project sufficient subfields in case of dereferences (i.e. removal of `y` from `B_ROW`)

Example:

transforms 
```
   Project(D := f1(A), E := f2(B_VARCHAR), G := f3(B_ROW.X))
       Unnest(replicate = [A], unnest = (B_ARRAY -> [B_BIGINT, B_VARCHAR, B_ROW]))
           Source(
               A bigint,
               B_ARRAY array(row(_bigint bigint, _varchar varchar, _row row(x bigint, y bigint))))
```

to 

```
   Project(D := f1(A), E := f2(B_VARCHAR), G := f3(B_ROW_X))
       Unnest(replicate = [A], unnest = (transformed -> [B_VARCHAR, B_ROW_X]))
           Project(A, transformed := transform(e -> row(e._varchar, e._row.x)))
               Source(A, B_ARAAY)
```

(Avoided creating two separate rules to prevent creating lambda expressions twice.)

Some special cases:

1) If the array element type is primitive, no pruning occurs.
2) If no fields are getting projected or used, we still keep at least one field in the mapping, since cardinality of the array is useful. We can replace it with a constant, but I'm not sure if the execution can be optimized to smartly create a ReplicateBlock for constant lambdas, and avoid copies.
3) If exactly one field is projected from the mapping, we do not wrap it with `ROW` construction, and instead kept as a primitive array.

the first two commits extend `ExpressionVerifier` to support verification of lambda expressions: https://github.com/prestosql/presto/pull/4270/commits/74a854c7cb9416e94e5a32b81768c497cd2d70d9#diff-51d1193486ad31af4d3dc949a1a9d401L511

### CODE REVIEW

File: presto-main/src/test/java/io/prestosql/sql/planner/assertions/ExpressionVerifier.java [hunk 1]
```diff
@@ -502,99 +521,154 @@ protected Boolean visitFunctionCall(FunctionCall actual, Node expectedExpression
...
+        private final Node expectedNode;
+        private final SymbolAliases symbolAliases;
+
+        Context(Node expectedNode, SymbolAliases symbolAliases)
+        {
+            this.expectedNode = expectedNode;
+            this.symbolAliases = symbolAliases;
+        }
+
+        Node getExpectedNode()
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/planner/assertions/ExpressionVerifier.java:659]: public

File: presto-main/src/test/java/io/prestosql/sql/planner/assertions/ExpressionVerifier.java [hunk 2]
```diff
@@ -502,99 +521,154 @@ protected Boolean visitFunctionCall(FunctionCall actual, Node expectedExpression
...
+        {
+            return expectedNode;
+        }
+
+        SymbolAliases getSymbolAliases()
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/planner/assertions/ExpressionVerifier.java:664]: public

File: presto-main/src/test/java/io/prestosql/sql/planner/assertions/ExpressionVerifier.java [hunk 3]
```diff
@@ -502,99 +521,154 @@ protected Boolean visitFunctionCall(FunctionCall actual, Node expectedExpression
...
+        {
+            return symbolAliases;
+        }
+
+        Context replaceExpectedNode(Expression expected)
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/planner/assertions/ExpressionVerifier.java:669]: public

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneUnnestMappings.java [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+ *          Source(
+ *              A bigint,
+ *              B_ARRAY array(row(_bigint bigint, _varchar varchar, _row row(x bigint, y bigint))))
+ *  </pre>
+ * to:
+ * <pre>
+ *  Project(D := f1(A), E := f2(B_VARCHAR), G := f3(B_ROW_X))
+ *      Unnest(replicate = [A], unnest = (transformed -> [B_VARCHAR, B_ROW_X]))
+ *          Project(A, transformed := transform(e -> row(e._varchar, e._row.x)))
+ *              Source(A, B_ARAAY)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneUnnestMappings.java:84]: Typo: B_ARAAY

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneUnnestMappings.java [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+
+        List<Symbol> newOutputSymbols = newOutputSymbolsBuilder.build();
+        List<Expression> newRowFields = newRowFieldsBuilder.build();
+
+        // Retain at least one output if everything is removed
+        // TODO: replace with a constant
+        if (newOutputSymbols.isEmpty()) {
+            newOutputSymbols = ImmutableList.of(outputSymbols.get(0));
+            newRowFields = ImmutableList.of(outputSymbols.get(0).toSymbolReference());
+        }
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneUnnestMappings.java:267]: Why do we need this?

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneUnnestMappings.java [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+        Symbol newInputSymbol = symbolAllocator.newSymbol(transformedArray, new ArrayType(prunedElementType));
+
+        return Optional.of(new Transformation(
+                new UnnestNode.Mapping(newInputSymbol, newOutputSymbols),
+                transformedArray,
+                dereferenceReplacements));
+    }
+
+    private LambdaExpression createLambdaExpression(
+            Symbol lambdaSymbol,
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneUnnestMappings.java:307]: `lambdaArgument`

File: presto-main/src/test/java/io/prestosql/sql/query/TestUnnest.java [hunk 7]
```diff
@@ -199,4 +199,17 @@ public void testRepeatedExpressions()
                 "SELECT * FROM (VALUES 1) t, UNNEST(ARRAY['a', 'b'], ARRAY['a', 'b']) u (x, y)"))
                 .matches("VALUES (1, 'a', 'a'), (1, 'b', 'b')");
     }
+
+    @Test
+    public void testDereferencesWithUnnest()
+    {
+        assertThat(assertions.query(
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/query/TestUnnest.java:206]: Can you add some more tests?

* One that references all unnested fields
* One that references some fields from both arrays
* One that references a whole row column (e.g., `z` in the query below)

**Review by martint — CHANGES_REQUESTED** [review]
The rule should not apply if all fields from a row are being referenced. E.g.,

```sql
WITH t(a) AS (
    VALUES (ARRAY[CAST(ROW(ROW(1, 2)) AS ROW(a ROW(f1 INTEGER, f2 INTEGER)))])
)
SELECT x.f1, x.f2
FROM t, UNNEST(t.a) t(x)
```

**Comment by colebow** [comment]
:wave: @phd3 - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.

**Pull request closed without merging**